clippy_lints/src/identity_op.rs

```diff
@@ -31,33 +32,99 @@ declare_clippy_lint! {
 declare_lint_pass!(IdentityOp => [IDENTITY_OP]);
 
 impl<'tcx> LateLintPass<'tcx> for IdentityOp {
-    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) {
-        if e.span.from_expansion() {
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &'tcx Body<'_>) {
+        if body.value.span.from_expansion() {
             return;
         }
-        if let ExprKind::Binary(cmp, left, right) = e.kind {
-            if is_allowed(cx, cmp, left, right) {
-                return;
+        let mut visitor = ExprVisitor {
+            cx,
+            leftmost: true,
+            top_level_binop: true,
+        };
+        visitor.visit_expr(&body.value);
+    }
+}
+
+struct ExprVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    leftmost: bool,
+    top_level_binop: bool,
+}
+
+impl<'a, 'tcx> ExprVisitor<'a, 'tcx> {
+    // Checks if `left op ..right` can be actually reduced to `right`
+    // e.g. `0 + if b { 1 } + else { 2 } +  if b { 3 } + else { 4 }`
+    // cannot be reduced to `if b { 1 } + else { 2 } +  if b { 3 } + else { 4 }`
+    // See #8724
+    fn reducible_to_right(&mut self, right: &'tcx Expr<'_>) -> bool {
+        if self.top_level_binop {
+            true
+        } else {
+            !if let ExprKind::If(..) | ExprKind::Match(..) = right.kind {
+                self.leftmost
+            } else {
+                false
             }
-            match cmp.node {
-                BinOpKind::Add | BinOpKind::BitOr | BinOpKind::BitXor => {
-                    check(cx, left, 0, e.span, right.span);
-                    check(cx, right, 0, e.span, left.span);
-                },
-                BinOpKind::Shl | BinOpKind::Shr | BinOpKind::Sub => check(cx, right, 0, e.span, left.span),
-                BinOpKind::Mul => {
-                    check(cx, left, 1, e.span, right.span);
-                    check(cx, right, 1, e.span, left.span);
-                },
-                BinOpKind::Div => check(cx, right, 1, e.span, left.span),
-                BinOpKind::BitAnd => {
-                    check(cx, left, -1, e.span, right.span);
-                    check(cx, right, -1, e.span, left.span);
-                },
-                BinOpKind::Rem => check_remainder(cx, left, right, e.span, left.span),
-                _ => (),
+        }
+    }
```

Comment:
Checks if `left op ...right` is reducible to `...right` based on the information gathered by traverser `ExprVisitor`.